tests/components/tplink/test_init.py

```diff
@@ -0,0 +1,169 @@
+async def test_configuring_discovery_disabled(hass):
+    """Test that discover does not get called when disabled."""
+    with mock_pyhs110, \
+            patch('homeassistant.components.tplink.async_setup_entry',
+                  return_value=mock_coro(True)) as mock_setup, \
+            patch('pyHS100.Discover.discover',
+                  return_value=[]) as discover:
+        await async_setup_component(hass, tplink.DOMAIN, {
+            tplink.DOMAIN: {
+                tplink.CONF_DISCOVERY: False
+            }
+        })
+        await hass.async_block_till_done()
+
+    assert len(discover.mock_calls) == 0
+    assert len(mock_setup.mock_calls) == 1
+
+
+async def test_platforms_are_initialized(hass):
+    """Test that platforms are initialized per configuration array."""
+    config = {
+        'tplink':
+            {'discovery': False,
+             'light': [{"host": "123.123.123.123"}],
+             'switch': [{"host": "321.321.321.321"}],
+             }
+    }
+
+    with patch('pyHS100.Discover.discover') as discover, \
+            patch('pyHS100.SmartDevice._query_helper'), \
+            patch('homeassistant.components.tplink.light.async_setup_entry',
+                  return_value=mock_coro(True)) as light_setup, \
+            patch('homeassistant.components.tplink.switch.async_setup_entry',
+                  return_value=mock_coro(True)) as switch_setup, \
+            patch('pyHS100.SmartPlug.is_dimmable', False):
+        # patching is_dimmable is necessray to avoid misdetection as light.
+        await async_setup_component(hass, tplink.DOMAIN, config)
+        await hass.async_block_till_done()
+
+    assert len(discover.mock_calls) == 0
+    assert len(light_setup.mock_calls) == 1
+    assert len(switch_setup.mock_calls) == 1
+
+
+async def test_no_config_creates_no_entry(hass):
+    """Test for when there is no tplink in config."""
+    with mock_pyhs110, \
+            patch('homeassistant.components.tplink.async_setup_entry',
+                  return_value=mock_coro(True)) as mock_setup:
+        await async_setup_component(hass, tplink.DOMAIN, {})
+        await hass.async_block_till_done()
+
+    assert len(mock_setup.mock_calls) == 0
+
+
+@pytest.mark.parametrize("platform", [
+    "switch",
+    "light",
+])
+async def test_unload(hass, platform):
+    """Test that the async_unload_entry works."""
+    # As we have currently no configuration, we just to pass the domain here.
+    entry = MockConfigEntry(domain=tplink.DOMAIN)
+    entry.add_to_hass(hass)
+
+    with patch('pyHS100.SmartDevice._query_helper'), \
+        patch('homeassistant.components.tplink.%s.async_setup_entry'
+              % platform,
+              return_value=mock_coro(True)) as light_setup:
+        config = {
+            'tplink': {
+                platform: [{"host": "123.123.123.123"}],
+                'discovery': False,
+            }
+        }
+        assert await async_setup_component(hass, tplink.DOMAIN, config)
```

Comment:
What happens when we already have a config entry added and try to setup the component? Will it try to create another config entry?

Reply:
To be honest, I'm not sure. Isn't that something the core is doing? I'm not explicitly doing anything special in the `setup_component` that I know of.

Reply:
`register_discovery_flow` only allows a single one 👍